shadowsocks/config.go

```diff
@@ -82,55 +204,104 @@ func ParseConfig(path string) (config *Config, err error) {
-		newField := newVal.Field(i)
-		oldField := oldVal.Field(i)
-		// log.Printf("%d: %s %s = %v\n", i,
-		// typeOfT.Field(i).Name, newField.Type(), newField.Interface())
-		switch newField.Kind() {
-		case reflect.Interface:
-			if fmt.Sprintf("%v", newField.Interface()) != "" {
-				oldField.Set(newField)
-			}
-		case reflect.String:
-			s := newField.String()
-			if s != "" {
-				oldField.SetString(s)
-			}
-		case reflect.Int:
-			i := newField.Int()
-			if i != 0 {
-				oldField.SetInt(i)
+
+	ProcessConfig(c)
+	return c, nil
+}
+
+func (c *Config) makeupServers() {
+	serversaddr := c.GetServerArray()
+	serversport := c.GetServerPortArray()
+	servers := make([]string, len(serversaddr))
+	for i := range servers {
+		servers[i] = net.JoinHostPort(serversaddr[i], serversport[i])
+	}
+	c.servers = servers
+}
+
+// ProcessConfig fill in the map after the config is unmarshaled
+func ProcessConfig(c *Config) {
+	c.PortPassword = make(map[string]string)
+	c.ServerPassword = make(map[string]string)
+
+	serversaddr := c.GetServerArray()
+	serversport := c.GetServerPortArray()
+	portspasswd := c.GetPasswordArray()
+
+	if len(serversaddr) != len(serversport) {
+		Logger.Fatal("error in config check, length of multi servers and ports mismatching")
+	}
+	if len(serversport) != len(portspasswd) {
+		Logger.Fatal("error in config check, length of multi ports and passwards mismatching")
+	}
+	c.makeupServers()
+
+	for i, v := range portspasswd {
+		c.PortPassword[serversport[i]] = v
+		c.ServerPassword[c.servers[i]] = v
+	}
+}
+
+// GetServer return the server array's first item
+func (c *Config) GetServer() string {
+	if len(c.servers) == 0 {
+		Logger.Fatal("error in get server, null slice returned")
+	}
+	return c.servers[0]
+}
+
+func (c *Config) GetServerRoundRobin() string {
+	defer func() { roundRobinIndex++ }()
+	serversaddr := c.GetServerArray()
+	serversport := c.GetServerPortArray()
+	index := roundRobinIndex % len(serversaddr)
+	return serversaddr[index] + ":" + serversport[index]
+}
+
+// Detect used only when multi tcp based ss server is setted for the client
+// Detect will try to dial each server to caculate a delay and sort server
+func (c *Config) Detect() {
+
+	type pair struct {
+		server string
+		delay  time.Duration
+	}
+
+	ping := func(addr string) time.Duration {
+		var avg time.Duration
+		for i := 0; i < 3; i++ {
+			t := time.Now()
+			_, err := net.Dial("tcp", addr)
+			ts := time.Now().Sub(t)
+			if err != nil {
+				return 0xfffffffffffffff
 			}
+			avg += ts
+		}
+		return avg / time.Duration(3)
+	}
+
+	var ss []pair
+	serversPort := c.GetServerPortArray()
+	for i, v := range c.GetServerArray() {
+		t := ping(v + ":" + serversPort[i])
+		ss = append(ss, pair{v, t})
+	}
+	sort.Slice(ss, func(i, j int) bool { return ss[i].delay < ss[j].delay })
+
+	var sortedserver string
+	for i, v := range ss {
+		if i == 0 {
+			sortedserver += v.server
 		}
+		sortedserver += ","
+		sortedserver += v.server
 	}
+	c.Server = sortedserver
```

Comment:
you should store the sorted server list into c.server,
because the       func (c *Config) GetServer() string         return server from it

Reply:
yes, right

Reply:
@dszhengyu fixed in 8796767